app/models/foreman_tasks/recurring_logic.rb

```diff
@@ -0,0 +1,53 @@
+module ForemanTasks
+
+  require 'parse-cron'
+
+  class RecurringLogic < ActiveRecord::Base
+    include Authorizable
+
+    has_one :task_group,
+            :class_name => TaskGroups::RecurringLogicTaskGroup.name
+
+    delegate :tasks, :to => :task_group
+
+    scoped_search :on => :id, :complete_value => false
+    scoped_search :on => :max_iteration, :complete_value => false, :rename => :iteration_limit
+    scoped_search :on => :iteration, :complete_value => false
+    scoped_search :on => :cronline, :complete_value => true
+
+    def next_occurrence_time(time = Time.now)
+      @parser ||= CronParser.new(cron_line)
+      @parser.next(time)
+    end
+
+    def generate_delay_options(time = Time.now, options = {})
+      {
+        :start_at => next_occurrence_time(time),
+        :start_before => options['start_before']
+      }
+    end
+
+    def can_continue?(time = Time.now)
+      (end_time.nil? || next_occurrence_time(time) < end_time) &&
+        (max_iteration.nil? || iteration < max_iteration)
+    end
+
+    def finished?
+      !can_continue? && task_group.tasks.all? { |task| %w(stopped paused).include? task.state }
+    end
+
+    def cancel_tasks
+      tasks.reject { |task| task.state != 'stopped' }.each { |task| task.cancel }
```

Comment:
using scope would be more efective, the `active` scope shold be the best fit, as sending cancel event to paused tasks wouldn't do anything anyway https://github.com/theforeman/foreman-tasks/blob/master/app/models/foreman_tasks/task.rb#L35